### PR TITLE
fix: Focus dispatching when token is removed

### PR DESCRIPTION
**Pull request by pan-kot**
### Description

Added explicit focus transitions when token group tokens are removed. The focus is transitioned to the next/prev available non-disabled items or, in a rare case, to the "show more" toggle.

Waiting for https://github.com/cloudscape-design/components/pull/876 to be merged first.

### How has this been tested?

New unit tests.

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/token-group/generic-token-group.tsx [hunk 1]
```diff
@@ -50,19 +53,69 @@ function GenericTokenGroup<Item>(
   const baseProps = getBaseProps(props);
   const className = clsx(baseProps.className, styles.root, hasItems && styles['has-items']);
 
+  useEffect(() => {
```
Comment (gethinwebster) [inline src/token-group/generic-token-group.tsx:56]: Presumably this is in an effect so that it's delayed until after the surrounding application has updated the state as a result of `onDismiss`? Can we reasonably rely on that? It seems like it's making quite a few assumptions about React internals and the surrounding application.

Comment (pan-kot) [reply src/token-group/generic-token-group.tsx:56]: It makes the following assumptions:

1. onDismiss removes the target item and and doesn't add, remove or change position of other items;
2. onDismiss is happening with a state update w/o an extra async delay.

The first assumption is something we don't expect to be violated. The second can be violated e.g. if the update would be wrapped in a `setTimeout`. In that case the focus would remain on the current item and then transition to the next focusable element - same as it was before. If we ever have this situation the only way to fix that would be to expose the token refs. However, the current solution would still work as a reasonable default.

Comment (gethinwebster) [reply src/token-group/generic-token-group.tsx:56]: ok, I think this makes sense. Might even steal this approach for some applayout focus handling fixes...

File: src/token-group/__tests__/token-group.test.tsx [hunk 2]
```diff
@@ -169,4 +185,64 @@ describe('TokenGroup', () => {
...
+    test('Focus is dispatched to the next active token when non-last token is removed', () => {
+      const wrapper = renderStatefulTokenGroup({
+        items: [
+          { label: '1', dismissLabel: 'Remove 1' },
+          { label: '2', dismissLabel: 'Remove 2' },
+          { label: '3', dismissLabel: 'Remove 3', disabled: true },
+          { label: '4', dismissLabel: 'Remove 4' },
+        ],
+      });
+      act(() => wrapper.findToken(2)!.findDismiss().click());
```
Comment (just-boris) [inline src/token-group/__tests__/token-group.test.tsx:199]: This `act` does nothing, after https://github.com/cloudscape-design/test-utils/pull/25

Why did you add it?

File: src/multiselect/__integ__/multiselect.test.ts [hunk 3]
```diff
@@ -108,7 +108,7 @@ const createSetupTest = (wrapper: MultiselectWrapper) => (testFn: (page: Multise
 });
 
 describe('focus handling', () => {
-  test('focuses trigger after dismissing a token', () => {
+  test('focuses trigger after dismissing all tokens', () => {
```
Comment (gethinwebster) [inline src/multiselect/__integ__/multiselect.test.ts:111]: do we really want to change this behavior? do we know why we had the old behavior in the first place?

File: src/token-group/__tests__/token-group.test.tsx (hunk 2 above)
Comment (pan-kot) [reply src/token-group/__tests__/token-group.test.tsx:199]: Thanks, Boris - I'll check.

File: src/multiselect/__integ__/multiselect.test.ts (hunk 3 above)
Comment (pan-kot) [reply src/multiselect/__integ__/multiselect.test.ts:111]: That looks reasonable to me that the tokens focusing behaviour is consistent no matter where the token group is used. However, I'll check if we do want a different treatment for multiselect.

Comment (gethinwebster) [reply src/multiselect/__integ__/multiselect.test.ts:111]: so long as someone on the design team has seen it it should be fine

**Review by just-boris — DISMISSED** [review]

File: src/token-group/__tests__/token-group.test.tsx [hunk 4]
```diff
@@ -169,4 +185,64 @@ describe('TokenGroup', () => {
...
+        items: [
+          { label: '1', dismissLabel: 'Remove 1' },
+          { label: '2', dismissLabel: 'Remove 2' },
+          { label: '3', dismissLabel: 'Remove 3', disabled: true },
+          { label: '4', dismissLabel: 'Remove 4' },
+        ],
+      });
+      wrapper.findToken(2)!.findDismiss().click();
+
+      expect(wrapper.findToken(3)!.findDismiss().getElement()).toHaveFocus();
```
[the remark this answers is not in the document]
Comment (pan-kot) [reply src/token-group/__tests__/token-group.test.tsx:201]: I am preferring to use indices here specifically because the purpose of the test is to check if the correct next/prev element is referenced so the indices convey it better I think.

But I will consider having both assertions for better regression.

File: src/token-group/generic-token-group.tsx [hunk 5]
```diff
@@ -19,8 +19,8 @@ interface ItemAttributes {
   disabled?: boolean;
   dismiss?: {
```
[the remark this answers is not in the document]
Comment (just-boris) [reply src/token-group/generic-token-group.tsx:20]: I would suggest this API

```js
<GenericToken
   {...itemAttributes}
   dissmiss={itemAttributes.dismiss && <DismissButton ref={ref} {...itemAttributes.dismiss} />}
/>
```

This way you also skip one level of ref forwarding.

File: src/multiselect/__integ__/multiselect.test.ts (hunk 3 above)
Comment (ethangeyer) [reply src/multiselect/__integ__/multiselect.test.ts:111]: UX approved :shipit:

**Review by gethinwebster — APPROVED** [review]

**Review by gethinwebster — APPROVED** [review]
